### PR TITLE
 feat: implement Partitioned DML transactions.

### LINKED ISSUE #287: Implement Transaction::PartitionedDmlOptions
We also need to add an overload to `Client::ExecuteSql` to accept an argument of type `Transaction::PartitionedDmlOptions` (just like how SingleUseOptions works today).

NOTE: there are no actual options today, but there might be some someday.

### LINKED ISSUE #432: Implement integration test for Client::ExecuteSql() with PartitionedDmlOptions

### PR DESCRIPTION
**Pull request by coryan**
Add `ExecuteSql()` overload that takes `Transaction::PartitionDmlOptions`.
Also add a unit test (only the success case), an integration test, and a couple
of examples. Applications cannot create a transaction object with these new
options because only one DML statement can be used with such transactions.

This fixes #287 and fixes #432.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/489)

### CODE REVIEW

File: google/cloud/spanner/transaction.h [hunk 1]
```diff
@@ -95,6 +97,23 @@ class Transaction {
...
+  /**
+   * Options for Partitioned DML transactions.
+   *
+   * @see
+   * https://cloud.google.com/spanner/docs/transactions#partitioned_dml_transactions
+   *   for an overview of Partitioned DML transactions.
+   */
+  class PartitionDmlOptions {
+   public:
+    // There are currently no read-write options.
```
Comment (devbww) [inline google/cloud/spanner/transaction.h]: s/read-write/partitioned-dml/

File: google/cloud/spanner/transaction.h [hunk 2]
```diff
@@ -95,6 +97,23 @@ class Transaction {
...
   };
 
+  /**
+   * Options for Partitioned DML transactions.
+   *
+   * @see
+   * https://cloud.google.com/spanner/docs/transactions#partitioned_dml_transactions
+   *   for an overview of Partitioned DML transactions.
+   */
+  class PartitionDmlOptions {
```
Comment (devbww) [inline google/cloud/spanner/transaction.h]: > Apart from the shape of comments, the transaction.{h,cc} changes are identical to mine.

Oh ... except that I called this `Partition*ed*DmlOptions`.

File: google/cloud/spanner/transaction.h (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/transaction.h]: Fixed.

File: google/cloud/spanner/transaction.h (hunk 2 above)
Comment (coryan) [reply google/cloud/spanner/transaction.h]: Fixed.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/spanner/transaction.cc [hunk 3]
```diff
@@ -86,6 +95,12 @@ Transaction::Transaction(ReadWriteOptions opts) {
   impl_ = std::make_shared<internal::TransactionImpl>(std::move(selector));
 }
 
+Transaction::Transaction(PartitionedDmlOptions opts) {
+  google::spanner::v1::TransactionSelector selector;
+  *selector.mutable_begin() = MakeOpts(std::move(opts.pd_opts_));
```
Comment (devjgm) [inline google/cloud/spanner/transaction.cc]: I think we should consider keeping "partitioned dml" stuff out of the Transaction classess and the `transaction.*` files. Instead, I think we can simply set the TransactionSelector proto directly in the implementation in `connection_impl.cc`. This would make it so the only place "partitioned dml" is uttered in our user-facing code is in the one function `Client::ExecutePartitionedDml`, which does not even take a `Transaction` nor options. I think the rest of the plumbing provided by the `Transaction` class is not needed for PartitionedDml transactions because they are never exposed to users, are not re-usable, and cannot be committed.

Comment (coryan) [reply google/cloud/spanner/transaction.cc]: Done.

**Comment by coryan** [comment]
PTAL.

**Comment by coryan** [comment]
Rebased to pick up less flaky builds.

**Comment by devjgm** [comment]
FWIW, this LGTM. Ping @devbww for approval.

File: google/cloud/spanner/client.h [hunk 4]
```diff
@@ -256,6 +256,26 @@ class Client {
...
+   * Executes a Partitioned DML SQL query.
+   *
+   * @param statement the SQL statement to execute. Please see the
+   *     [spanner documentation][dml-partitioned] for the restrictions on the
+   *     SQL statements supported by this function.
+   *
+   * @par Example
+   * @snippet samples.cc execute-sql-partitioned
+   *
+   * @see [Partitioned DML Transactions][txn-partitioned] for an overview of
```
Comment (devbww) [inline google/cloud/spanner/client.h]: Do we need to mention "Partitioned DML Transactions" to the user now?

File: google/cloud/spanner/client.h [hunk 5]
```diff
@@ -256,6 +256,26 @@ class Client {
...
+   *     Partitioned DML transactions.
+   * @see [Partitioned DML][dml-partitioned] for a description of which SQL
```
Comment (devbww) [inline google/cloud/spanner/client.h]: Is this redundant with `@param statement`?

File: google/cloud/spanner/integration_tests/client_integration_test.cc [hunk 6]
```diff
@@ -341,6 +341,28 @@ TEST_F(ClientIntegrationTest, ExecuteSql) {
...
+                                          {"lname", Value("test-lname-1")}}));
+        if (!insert) return std::move(insert).status();
+        return Mutations{};
+      });
+  EXPECT_STATUS_OK(insert_result);
+
+  auto result = client_->ExecutePartitionedDml(
+      SqlStatement("UPDATE Singers SET LastName = 'test-only'"
+                   " WHERE SingerId >= 1"));
+  EXPECT_STATUS_OK(result);
```
Comment (devbww) [inline google/cloud/spanner/integration_tests/client_integration_test.cc]: Should we try to read this back, like the other mutating cases?

File: google/cloud/spanner/internal/partial_result_set_reader.cc [hunk 7]
```diff
@@ -83,10 +83,15 @@ StatusOr<optional<Value>> PartialResultSetReader::NextValue() {
 
 PartialResultSetReader::~PartialResultSetReader() {
   if (!finished_) {
+    // If there is actual data in the streaming RPC Finish() can deadlock, so
```
Comment (devbww) [inline google/cloud/spanner/internal/partial_result_set_reader.cc]: Should the edits in this file be in a separate PR?

File: google/cloud/spanner/partitioned_dml_result.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+/**
+ * The result of executing a Partitioned DML query.
+ */
+struct PartitionedDmlResult {
+  /// A lower bound on the number of rows modified.
+  std::int64_t row_count_lower_bound;
```
Comment (devbww) [inline google/cloud/spanner/partitioned_dml_result.h]: `#include <cstdint>`

File: google/cloud/spanner/partitioned_dml_result.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_PARTITIONED_DML_RESULT_H_
+#define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_PARTITIONED_DML_RESULT_H_
+
+#include "google/cloud/spanner/timestamp.h"
```
Comment (devbww) [inline google/cloud/spanner/partitioned_dml_result.h]: Unused.

File: google/cloud/spanner/samples/samples.cc [hunk 10]
```diff
@@ -391,6 +391,27 @@ void DmlStandardDelete(google::cloud::spanner::Client client) {
...
+  if (!result) throw std::runtime_error(result.status().message());
+  std::cout << "Delete was successful [spanner_dml_partitioned_delete]\n";
+}
+//! [execute-sql-partitioned] [END spanner_dml_partitioned_delete]
+
+//! [START spanner_dml_partitioned_update]
+void DmlPartitionedUpdate(google::cloud::spanner::Client client) {
+  namespace spanner = google::cloud::spanner;
+  auto result = client.ExecutePartitionedDml(
+      spanner::SqlStatement("UPDATE Albums SET MarketingBudget = 100000"
```
Comment (devbww) [inline google/cloud/spanner/samples/samples.cc]: How does this sample know `MarketingBudget` exists?

File: google/cloud/spanner/samples/samples.cc [hunk 11]
```diff
@@ -680,6 +703,16 @@ void RunAll() {
   std::cout << "\nRunning spanner_field_access_on_nested_struct sample\n";
   FieldAccessOnNestedStruct(client);
 
+  namespace s = google::cloud::spanner;
+  s::Client c2(s::MakeConnection(
+      s::Database(project_id, instance_id, database_id),
+      s::ConnectionOptions().enable_clog().enable_tracing("rpc")));
```
Comment (devbww) [inline google/cloud/spanner/samples/samples.cc]: Is this a "submit" thing or a "development testing" thing?

Comment (coryan) [reply google/cloud/spanner/samples/samples.cc]: Ooops, fixed, thanks!

File: google/cloud/spanner/samples/samples.cc (hunk 10 above)
Comment (coryan) [reply google/cloud/spanner/samples/samples.cc]: The same way all the examples know that the tables exist, or any other column. They need to be executed in a particular order.

File: google/cloud/spanner/partitioned_dml_result.h (hunk 9 above)
Comment (coryan) [reply google/cloud/spanner/partitioned_dml_result.h]: Removed.

File: google/cloud/spanner/partitioned_dml_result.h (hunk 8 above)
Comment (coryan) [reply google/cloud/spanner/partitioned_dml_result.h]: Added.

File: google/cloud/spanner/internal/partial_result_set_reader.cc (hunk 7 above)
Comment (coryan) [reply google/cloud/spanner/internal/partial_result_set_reader.cc]: Maybe... It hangs without them, so we need them before or with this PR.

File: google/cloud/spanner/integration_tests/client_integration_test.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/spanner/integration_tests/client_integration_test.cc]: Fixed.

File: google/cloud/spanner/client.h (hunk 5 above)
Comment (coryan) [reply google/cloud/spanner/client.h]: Removed.

File: google/cloud/spanner/client.h (hunk 4 above)
Comment (coryan) [reply google/cloud/spanner/client.h]: Removed.

File: google/cloud/spanner/samples/samples.cc [hunk 12]
```diff
@@ -703,15 +703,11 @@ void RunAll() {
   std::cout << "\nRunning spanner_field_access_on_nested_struct sample\n";
   FieldAccessOnNestedStruct(client);
 
-  namespace s = google::cloud::spanner;
-  s::Client c2(s::MakeConnection(
-      s::Database(project_id, instance_id, database_id),
-      s::ConnectionOptions().enable_clog().enable_tracing("rpc")));
-  std::cout << "\nRunning spanner_dml_partitioned_update sample\n";
-  DmlPartitionedUpdate(c2);
+  std::cout << "\nRunning spanner_dml_partitioned_update sample\n" << std::endl;
```
Comment (devbww) [inline google/cloud/spanner/samples/samples.cc]: `<< "\n" << std::endl` ?

File: google/cloud/spanner/samples/samples.cc [hunk 13]
```diff
@@ -703,15 +703,11 @@ void RunAll() {
...
+  DmlPartitionedUpdate(client);
 
   std::cout << "\nRunning spanner_dml_partitioned_delete sample\n" << std::endl;
```
Comment (devbww) [inline google/cloud/spanner/samples/samples.cc]: Ditto.

File: google/cloud/spanner/internal/partial_result_set_reader.cc (hunk 7 above)
Comment (devbww) [reply google/cloud/spanner/internal/partial_result_set_reader.cc]: OK.  I'm fine with them here, but let's tag @mr-salty so he knows.

File: google/cloud/spanner/samples/samples.cc (hunk 10 above)
Comment (devbww) [reply google/cloud/spanner/samples/samples.cc]: But `samples.cc` also has `RunOneCommand` support.

File: google/cloud/spanner/samples/samples.cc (hunk 12 above)
Comment (coryan) [reply google/cloud/spanner/samples/samples.cc]: Fixed.

File: google/cloud/spanner/samples/samples.cc (hunk 13 above)
Comment (coryan) [reply google/cloud/spanner/samples/samples.cc]: Fixed.

File: google/cloud/spanner/samples/samples.cc (hunk 10 above)
Comment (coryan) [reply google/cloud/spanner/samples/samples.cc]: Yes. And it was true for other existing examples that use `MarketingBudget`, and it is also true in the samples for other languages (and libraries).  I get the point, it means the samples can fail if you don't setup your environment "just so", but (I think) this is not the PR to fix that (assuming we want to fix it).

Comment (devbww) [reply google/cloud/spanner/samples/samples.cc]: Fair enough, but I wasn't suggesting this was the PR to fix it ... just the PR not to introduce another occurrence.

File: google/cloud/spanner/internal/connection_impl.cc [hunk 14]
```diff
@@ -265,6 +276,52 @@ StatusOr<ResultSet> ConnectionImpl::ExecuteSql(
   return ResultSet(std::move(*reader));
 }
 
+StatusOr<PartitionedDmlResult> ConnectionImpl::ExecutePartitionedDml(
+    SessionHolder& session, google::spanner::v1::TransactionSelector& s,
+    std::int64_t seqno, ExecutePartitionedDmlParams epdp) {
+  if (session.session_name().empty()) {
+    // Since the session may be sent to other machines, it should not be
+    // returned to the pool when the Transaction is destroyed (release=true).
+    auto session_or = GetSession(/*release=*/true);
```
Comment (mr-salty) [inline google/cloud/spanner/internal/connection_impl.cc]: I think this is a bug - PartitionedDml transactions are not partitioned among clients like the other Partition* cases, right?  @scotthart can you confirm?

Comment (mr-salty) [reply google/cloud/spanner/internal/connection_impl.cc]: FYI I sent #563 - we can continue the discussion there.
